### PR TITLE
SLES4SAP PINT query function

### PR DESCRIPTION
**Pull request by mpagot**
MU jobs for SAP and HA tests in the cloud are designed to:
- create a HA or SAP cluster in the cloud
- install the update on it
- perform some functional tests.

The cluster is created using various versions of SLES images from the CSP catalog.
This change is about always providing the openQA POST the name of the latest version of the catalog image.
This information is retrieved from PINT
 
Content of the PR:
 - Add API to PC helper to query could imaged from PINT.
 - Add Unit Testing coverage for that.
 - Add example of usage in pc_helper_online.

Ticket TEAM-8180

### CODE REVIEW

File: openqabot/pc_helper.py [hunk 1]
```diff
@@ -100,6 +100,148 @@ def apply_publiccloud_pint_image(settings):
...
+        job_settings["SLES4SAP_QESAP_OS_VER_REGIONS"] = ";".join(setting_regions)
+        job_settings["SLES4SAP_QESAP_OS_VER_ID"] = ";".join(setting_ami)
+    return job_settings
+
+
+def apply_sles4sap_pint_image(
+    cloud_provider, pint_base_url, name_filter, region_list=None
+):
+    """
+    Applies OS_IMAGE relates settings based on the given SLES4SAP_IMAGE_REGEX
```
Comment (alvarocarvajald) [inline openqabot/pc_helper.py:227]: s/relates/related/

File: pc_helper_online.py [hunk 2]
```diff
@@ -47,6 +51,23 @@ def main():
...
+                if "SLES4SAP_PINT_QUERY" in settings and "incidents" in data:
+                    for flavor in data["incidents"]["FLAVOR"]:
+                        if (
+                            "pint_name_regexp"
+                            not in data["incidents"]["FLAVOR"][flavor]
+                        ):
+                            continue
+                        cloud_provider = ""
+                        if "azure" in flavor.lower():
+                            cloud_provider = "AZURE"
```
Comment (alvarocarvajald) [inline pc_helper_online.py:63]: Why not
```
cloud_provider = flavor.lower()
```
Instead of these 3 lines? Do we need `cloud_provider` to be empty on GCE and EC2?

Comment (mpagot) [reply pc_helper_online.py:63]: To be more flexible

TL;DR
I'm not even sure that inspecting FLAVOR is a good idea at all. Problem is that we do not have any strongly enforced policy in the sles4sap metadata files about the flavor names for cloud build. Usually we fit in there the cloud provider name, we fit the BYOS/PAYG. So I decided to use this more generic test criteria to be more flexible. Another possible approach is not to use the FLAVOR at all and add an explicit.

**Review by alvarocarvajald — APPROVED** [review]
LGTM

**Review by lpalovsky — APPROVED** [review]
LGTM

**Review by okurz — CHANGES_REQUESTED** [review]
This is adding quite a lot of changes. Could you please share more context about this, what benefit it brings, what it is required for, etc. The "TEAM-XXX" link does not help as it's not publically available.

@asmorodskyi as an important recent contributor please be aware about this change and potentially review

File: openqabot/pc_helper.py [hunk 3]
```diff
@@ -100,6 +102,152 @@ def apply_publiccloud_pint_image(settings):
...
+    :param str region optional filter, mostly needed by AWS
+    :return Return dictionary has keys NAME and ID
+    """
+    settings = {}
+    try:
+        url = f"{query}{state}.json"
+        log.debug("Pint url:%s name_filter:%s region:%s", url, name_filter, region)
+        images = pint_query(url)["images"]
+        image = get_recent_pint_image(images, name_filter, region=region, state=state)
+        if image is None:
```
Comment (asmorodskyi) [inline openqabot/pc_helper.py:126]: mmm will BaseException just catch error which you throwing ? why not just do return settings ?

**Comment by asmorodskyi** [comment]
why do you need PINT for Azure and Google ? In Azure there is nice mechanism of "offer" and "sku" . You can just say Azure "give me latest SLE 15 SP2 image" and Azure will handle what you trying to do here in bot . Something similar about Google . There is good reason why we have such logic only in AWS because only AWS API refusing to "speak" with us using human understandable SLE versions and requires some wicked UUID's which only PINT knows but this is not the case with Azure or Google

**Comment by mpagot** [comment]
> This is adding quite a lot of changes. Could you please share more context about this, what benefit it brings, what it is required for, etc. The "TEAM-XXX" link does not help as it's not publically available.

PR description extended with more context

**Comment by mpagot** [comment]
> why do you need PINT for Azure and Google ? In Azure there is nice mechanism of "offer" and "sku" . You can just say Azure "give me latest SLE 15 SP2 image" and Azure will handle what you trying to do here in bot .

true, I can potentially simplify the PR

**Comment by mpagot** [comment]
> Something similar about Google .

Do you have an example of the syntax to be used?

**Comment by alvarocarvajald** [comment]
@asmorodskyi wrote:
> why do you need PINT for Azure and Google ?\

+1 on the Azure front as we can already reference the latest image with `:latest` in the offer (IIRC), but is there already such a mechanism for Google?

As far as we understood, Google images have to be referenced by project and image name (for example: `sles-12-sp5-sap-v20230718-x86-64`), and the image name contains a date.

File: openqabot/pc_helper.py [hunk 4]
```diff
@@ -100,6 +102,133 @@ def apply_publiccloud_pint_image(settings):
...
+    job_settings = {}
+    ret = get_pint_image(
+        name_filter=name_filter,
+        field="project",
+        state=state,
+        query=pint_url(pint_base_url, "google"),
+    )
+    if any(ret):
+        job_settings["SLES4SAP_QESAP_OS_VER"] = f"{ret['ID']}/{ret['NAME']}"
+        job_settings["SLES4SAP_QESAP_OS_STATE"] = state
```
Comment (mpagot) [inline openqabot/pc_helper.py:166]: @asmorodskyi yesterday we always talk about the fact that dumping `state` to a setting is not needed, but PC is dumping it too

https://github.com/openSUSE/qem-bot/pull/144/files#diff-5578fec6c7adac2cb35f299569fc8310e4a4e7d1005ac8d1802d3dc75b20ede1R84

seems nothing on the test side it using `PUBLIC_CLOUD_IMAGE_STATE`. Would you like me to remove `PUBLIC_CLOUD_IMAGE_STATE` too?

File: openqabot/pc_helper.py [hunk 5]
```diff
@@ -100,6 +102,133 @@ def apply_publiccloud_pint_image(settings):
...
+            region=this_region,
+        )
+        if any(ret):
+            images_list[this_region] = ret
+    if any(images_list):
+        # All element should have same name, just get the first
+        job_settings["SLES4SAP_QESAP_OS_VER"] = next(iter(images_list.items()))[1][
+            "NAME"
+        ]
+        job_settings["SLES4SAP_QESAP_OS_VER_STATE"] = state
```
Comment (mpagot) [inline openqabot/pc_helper.py:194]: At least I should rename this one from `SLES4SAP_QESAP_OS_VER_STATE` to `SLES4SAP_QESAP_OS_STATE`

File: openqabot/pc_helper.py [hunk 6]
```diff
@@ -100,6 +102,133 @@ def apply_publiccloud_pint_image(settings):
...
+    """
+    job_settings = {}
+    ret = get_pint_image(
+        name_filter=name_filter,
+        field="project",
+        state=state,
+        query=pint_url(pint_base_url, "google"),
+    )
+    if any(ret):
+        job_settings["SLES4SAP_QESAP_OS_VER"] = f"{ret['ID']}/{ret['NAME']}"
```
Comment (mpagot) [inline openqabot/pc_helper.py:165]: I can potentially rename this one to `PUBLIC_CLOUD_IMAGE_NAME`. I'd prefer to do it later as having a dedicated name could help me to debug the final result looking at the triggered job settings.

File: openqabot/pc_helper.py [hunk 7]
```diff
@@ -100,6 +102,131 @@ def apply_publiccloud_pint_image(settings):
...
+            raise ValueError(
+                f"Cannot find matching image in PINT with name:[{name_filter}] and state:{state}"
+            )
+        if field not in image.keys() or "name" not in image.keys():
+            raise ValueError(
+                f"Cannot find expected keys '{field}' in the selected image dictionary {image}"
+            )
+        settings["ID"] = image[field]
+        settings["NAME"] = image["name"]
+    except BaseException as e:
```
Comment (ricardobranco777) [inline openqabot/pc_helper.py:136]:
```suggestion
    except BaseException as e:
```
```suggestion
    except Exception as e:
```

File: openqabot/pc_helper.py [hunk 8]
```diff
@@ -100,6 +102,131 @@ def apply_publiccloud_pint_image(settings):
...
+    return settings
+
+
+def pint_url(pint_base_url, csp_name):
+    """
+    :param str pint_base_url, usually https://susepubliccloudinfo.suse.com/v1
+    :param str csp_name microsoft, amazon, google, ...
+    :return str PINT url for the images API
+    """
+    url = "/".join([pint_base_url] + [csp_name, "images"]) + "/"
```
Comment (ricardobranco777) [inline openqabot/pc_helper.py:148]:
```suggestion
    url = f"{pint_base_url}/{csp_name}/images/"
```

File: openqabot/pc_helper.py [hunk 9]
```diff
@@ -100,6 +102,131 @@ def apply_publiccloud_pint_image(settings):
...
+    Query PINT about GCE images and retrieve the latest one
+    """
+    job_settings = {}
+    ret = get_pint_image(
+        name_filter=name_filter,
+        field="project",
+        state=state,
+        query=pint_url(pint_base_url, "google"),
+    )
+    if any(ret):
```
Comment (ricardobranco777) [inline openqabot/pc_helper.py:164]: Why use `any` here?

File: openqabot/pc_helper.py [hunk 10]
```diff
@@ -100,6 +102,131 @@ def apply_publiccloud_pint_image(settings):
...
+    images_list = {}
+    for this_region in region_list:
+        ret = get_pint_image(
+            name_filter=name_filter,
+            field="id",
+            state=state,
+            query=pint_url(pint_base_url, "amazon"),
+            region=this_region,
+        )
+        if any(ret):
```
Comment (ricardobranco777) [inline openqabot/pc_helper.py:186]: Why use any() on a dict?

File: openqabot/pc_helper.py (hunk 6 above)
Comment (pdostal) [reply openqabot/pc_helper.py:165]: I don't understand how this can benefit. I only see yet another unnecessary variable.

File: openqabot/pc_helper.py (hunk 4 above)
Comment (asmorodskyi) [reply openqabot/pc_helper.py]: no we exposing it for informational purposes . So for example when you see job for SLE 15 SP5 which running with image in "deprecated" state you suppose to figure out why this is happening because SP5 which is latest currently suppose to have "active" images , so there is issue or with bot or with image publishing pipeline

**Review by pdostal — CHANGES_REQUESTED** [review]
I don't see a single point why you do this.

All of this logic is already done. You just create your own metadata files.
If there is something different then just alter existing code please.

File: openqabot/pc_helper.py (hunk 5 above)
Comment (asmorodskyi) [reply openqabot/pc_helper.py]: why can't you use `PUBLIC_CLOUD_IMAGE_STATE` ?

File: openqabot/pc_helper.py (hunk 3 above)
Comment (mpagot) [reply openqabot/pc_helper.py:126]: in order to return to the user a more meaningful error message

**Comment by a-kpappas** [comment]
@pdostal How would you 
>  provide the openQA POST the name of the latest version of the catalog image.
> This information is retrieved from PINT
?

File: openqabot/pc_helper.py (hunk 6 above)
Comment (mpagot) [reply openqabot/pc_helper.py:165]: renaming is one of the suggestion that I get from Anton yesterday

**Comment by a-kpappas** [comment]
I added non-qesap members as stakeholders to the ticket. You should have visibility now.

File: openqabot/pc_helper.py (hunk 5 above)
Comment (mpagot) [reply openqabot/pc_helper.py]: is it `state` needed by the test code? I'm not able to find any occurrence of `PUBLIC_CLOUD_IMAGE_STATE` in https://github.com/os-autoinst/os-autoinst-distri-opensuse

File: openqabot/pc_helper.py (hunk 4 above)
Comment (mpagot) [reply openqabot/pc_helper.py]: I like the idea, I'll do it for sles4sap and reusing the `PUBLIC_CLOUD_IMAGE_STATE` name

**Comment by pdostal** [comment]
So from my understanding this pull request does the following:
 * Move control over which region to test level:
   * Why? - It can be done in metadata.
   * In metadata every flavour just needs to be in separate file but it's anyways quite practical and easy to maintain.
 * Return multiple AMIs of multiple regions so each test suite can run in different region:
   * This is doable but needs to be better documented and of course paired with pull request to os-autoinst-distri-opensuse.

File: openqabot/pc_helper.py (hunk 3 above)
Comment (asmorodskyi) [reply openqabot/pc_helper.py:126]: ok

**Comment by asmorodskyi** [comment]
> As far as we understood, Google images have to be referenced by project and image name (for example: `sles-12-sp5-sap-v20230718-x86-64`), and the image name contains a date.
> Yes, an example would help a lot.

sorry to confuse you , you totally right here , for Google we also using pint to figure out exact image name because it contains date which is always changing ...

File: openqabot/pc_helper.py (hunk 8 above)
Comment (m-dati) [reply openqabot/pc_helper.py:148]: I'd suggest to remove last **`'/'`**, in order to use then:
`url/<something else>`
instead of:
`ulr<something else> `less clear.

File: openqabot/pc_helper.py (hunk 7 above)
Comment (mpagot) [reply openqabot/pc_helper.py:136]: @ricardobranco777  could you please elaborate about why `Exception` is better?

File: openqabot/pc_helper.py (hunk 8 above)
Comment (mpagot) [reply openqabot/pc_helper.py]: done https://github.com/openSUSE/qem-bot/compare/f086781efbf8ac1d8e0528b93812a4d16e801136..9693a7e55323377acd5daa10a80b943dc07f48f5

File: pc_helper_online.py [hunk 11]
```diff
@@ -47,6 +51,23 @@ def main():
                     apply_publiccloud_pint_image(settings)
                     if "PUBLIC_CLOUD_IMAGE_ID" not in settings:
                         log.error(f"Failed to get PUBLIC_CLOUD_IMAGE_ID from {data}")
+                if "SLES4SAP_PINT_QUERY" in settings and "incidents" in data:
+                    for flavor in data["incidents"]["FLAVOR"]:
+                        if (
+                            "pint_name_regexp"
```
Comment (mpagot) [inline pc_helper_online.py:57]: @asmorodskyi  @pdostal this is only partially implemented but relevant clue about how next PR will handle metadata. As reference here a metadata example:

```
 product: SAP-15-SP5
   2   │ settings:
   3   │   DISTRI: sle
   4   │   VERSION: '15-SP5'
   5 + │   SLES4SAP_PINT_QUERY: 'https://susepubliccloudinfo.suse.com/v1'
   6   │ aggregate:
...
  18   │ incidents:
  19   │   FLAVOR:
...
  65   │     Azure-SAP-BYOS-Incidents-saptune:
  66   │       aggregate_job: false
  67   │       archs:
  68   │         - x86_64
  69   │       issues: *test_issues_list
  70   │       required_issues: *required_list
  71   │       packages: *whitelist_saptune
  72 + │       pint_name_regexp: '.*sap-15-sp5-byos'

```

Comment (pdostal) [reply pc_helper_online.py:57]: Why `SLES4SAP_PINT_QUERY` and not `PUBLIC_CLOUD_PINT_QUERY`?

Comment (mpagot) [reply pc_helper_online.py:57]: In order to call `apply_sles4sap_pint_image` in place of `apply_publiccloud_pint_image`
